src/expr.rs

```diff
@@ -19,194 +19,246 @@ use syntax::print::pprust;
+    };
 
-        format!("{}({})", callee_str, args_str)
-    }
+    Some(format!("{}({})", callee_str, args_str))
+}
 
-    fn rewrite_paren(&mut self, subexpr: &ast::Expr, width: usize, offset: usize) -> String {
-        debug!("rewrite_paren, width: {}, offset: {}", width, offset);
-        // 1 is for opening paren, 2 is for opening+closing, we want to keep the closing
-        // paren on the same line as the subexpr
-        let subexpr_str = self.rewrite_expr(subexpr, width-2, offset+1);
-        debug!("rewrite_paren, subexpr_str: `{}`", subexpr_str);
-        format!("({})", subexpr_str)
-    }
+fn rewrite_paren(context: &RewriteContext, subexpr: &ast::Expr, width: usize, offset: usize) -> Option<String> {
+    debug!("rewrite_paren, width: {}, offset: {}", width, offset);
+    // 1 is for opening paren, 2 is for opening+closing, we want to keep the closing
+    // paren on the same line as the subexpr
+    let subexpr_str = subexpr.rewrite(context, width-2, offset+1);
+    debug!("rewrite_paren, subexpr_str: `{:?}`", subexpr_str);
+    subexpr_str.map(|s| format!("({})", s))
+}
 
-    fn rewrite_struct_lit(&mut self,
+fn rewrite_struct_lit(context: &RewriteContext,
                           path: &ast::Path,
                           fields: &[ast::Field],
                           base: Option<&ast::Expr>,
                           width: usize,
                           offset: usize)
-        -> String
-    {
-        debug!("rewrite_struct_lit: width {}, offset {}", width, offset);
-        assert!(fields.len() > 0 || base.is_some());
+        -> Option<String>
+{
+    debug!("rewrite_struct_lit: width {}, offset {}", width, offset);
+    assert!(fields.len() > 0 || base.is_some());
 
-        let path_str = pprust::path_to_string(path);
-        // Foo { a: Foo } - indent is +3, width is -5.
-        let indent = offset + path_str.len() + 3;
-        let budget = width - (path_str.len() + 5);
+    let path_str = pprust::path_to_string(path);
+    // Foo { a: Foo } - indent is +3, width is -5.
+    let indent = offset + path_str.len() + 3;
+    let budget = width - (path_str.len() + 5);
 
-        let mut field_strs: Vec<_> =
-            fields.iter().map(|f| self.rewrite_field(f, budget, indent)).collect();
-        if let Some(expr) = base {
-            // Another 2 on the width/indent for the ..
-            field_strs.push(format!("..{}", self.rewrite_expr(expr, budget - 2, indent + 2)))
+    let mut field_strs = Vec::with_capacity(fields.len());
+    for field in fields.iter() {
+        match rewrite_field(context, field, budget, indent) {
+            Some(s) => { field_strs.push(s); }
+            None => { return None; }
         }
+    }
+    if let Some(expr) = base {
+        // Another 2 on the width/indent for the ..
+        field_strs.push(match expr.rewrite(context, budget - 2, indent + 2) {
+            Some(s) => format!("..{}", s),
+            None => { return None; }
+        });
```

Comment:
We should use a macro for this pattern - it is just an Option's equivalent of `try!`, but would save a tonne of code.